aws/resource_aws_s3_bucket_object.go

```diff
@@ -269,6 +301,100 @@ func resourceAwsS3BucketObjectPut(d *schema.ResourceData, meta interface{}) erro
 	return resourceAwsS3BucketObjectRead(d, meta)
 }
 
+func resourceAwsS3BucketObjectCopy(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	restricted := meta.(*AWSClient).IsChinaCloud()
+
+	destBucket := d.Get("bucket").(string)
+	destKey := d.Get("key").(string)
+
+	copySource := strings.TrimLeft(d.Get("source").(string), "s3://")
+	if !strings.Contains(copySource, "/") {
+		return fmt.Errorf(
+			"S3 source (%s) must be the name of the source bucket and key name of the source object, separated by a slash (/)",
+			copySource)
+	}
+	copySource = url.QueryEscape(fmt.Sprintf("/%s", copySource))
+
+	// data structure
+	copyInput := &s3.CopyObjectInput{
+		Bucket:     aws.String(destBucket),
+		Key:        aws.String(destKey),
+		CopySource: aws.String(copySource),
+	}
+
+	// replace with settings if they exist
+	if v, ok := d.GetOk("acl"); ok {
+		copyInput.ACL = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("storage_class"); ok {
+		copyInput.StorageClass = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("cache_control"); ok {
+		copyInput.CacheControl = aws.String(v.(string))
```

Comment:
In the copy context, this seems to be of no effect if the original object has cache control.